renderer/src/main.tsx

```diff
@@ -14,3 +14,11 @@ ReactDOM.createRoot(
     </BrowserRouter>
   </React.StrictMode>
 )
+
+window.electron.onActivityLogged(entry => {
+  console.log('[ACTIVITY] %j', entry)
+})
+
+window.electron.resumeActivityStream().then(() => {
+  console.log('ACTIVITY STREAM RESUMED')
+})
```

Comment:
Please take a look at this example usage of the communication channel between the backend and the frontend for transmitting Activity Log entries. 

Is this new API (`onActivityLogged` and `resumeActivityStream`) easy enough to consume from a React frontend?

Are there any improvements we can make to make the integration with React simpler?

Reply:
Would it be helpful if I wrote down some documentation for this API, perhaps as JSDoc comments in `renderer/src/typings.d.ts` below?

Reply:
- `onActivityLogged`: React hooks will want a way to unsubscribe as well, would it be simpler to let react register and unregister its own event listeners on the `ipc` object? Or if you still want to abstract it away, we could forward the events to a new event emitter, like this:


```js
window.electron.events.on('activity', () => {})
window.electron.events.off(...)
```

Could also call it `window.electron.activityLog`.

- `resumeActivityStream()`: This should work well as a `useEffect()`, however I'm also concerned about the clean up, and I think this method will need a matching `stopActivityStream()`.

What do you think about these function names, to make more clear what's happening?

```js
useEffect(() => {
  window.electron.activityLog.on('event', onEvent)
  window.electron.activityLog.subscribe()
  return () => {
    window.electron.activityLog.unsubscribe()
    window.electron.activityLog.removeEventListener('event', onEvent)
  }
})
```

Now that I read it, this pair of functions is awkward, shall we diverge from the EventEmitter pattern and implement it somewhat like this?

```js
useEffect(() => {
  const unsubscribe = window.electron.activityLog.subscribe(onEvent)
  return unsubscribe
})
```

Reply:
I am not sure if it's a good idea to subscribe & unsubscribe from inside `useEffect`. 

I know very little about React hooks, so I may be wrong, but my concern is that if React decides to re-run our component, and inside `useEffect` decides to create a new "hook instance", i.e. calls `unsubscribe` from the last instance and then creates a new instance by calling our `useEffect` callback, then we have a race condition. If a new activity is reported in the window between `unsubscribe` and `subscribe`, the we will not record it. Again, I am not sure if this is a possible scenario with React hooks.

IMO, we need the subscription to be long-lived, and keep the list of all Activity entries received so far in a persistent array (a global state).

I would think about this subscription as similar to a web socket connection. We don't want to subscribe/unsubscribe for web socket events from a component.

While searching for a solution, I found this article describing a solution based on `React.createContext`, which seems very reasonable to my beginner eyes.
https://aravindballa.com/writings/custom-hook-to-listen-websockets/

Reply:
I was thinking about implementing an `ActivityLogProvider` to store any long-living state & listeners, plus `useActivityLogSubscribe()` using `useContext` and `useEffect` for consumption by the actual app components.

```tsx
const App = () => (
  <ActivityLogProvider >
    {/* the actual app */}
    <ExampleComponent />
  </ActivityLogProvider >
);
```

Reply:
> If a new activity is reported in the window between `unsubscribe` and `subscribe`, the we will not record it. Again, I am not sure if this is a possible scenario with React hooks.

As I was thinking about this, I realized there won't be any race condition. Instead, when we subscribe after unsubscribing, the backend will re-emit all existing ActivityLog entries for us. So no data is lost, but it makes me wonder what will be the impact on performance.

Reply:
Setting my concerns aside, I really like your proposed API 👍🏻 

```ts
useEffect(() => {
  const unsubscribe = window.electron.activityLog.subscribe(onEvent)
  return unsubscribe
})
```

I would even go one step further (or maybe backward towards my current API):

```ts
useEffect(() => {
  const unsubscribe = window.electron.subscribeToActivityLog(onEvent)
  return unsubscribe
})
```

Reply:
> So no data is lost, but it makes me wonder what will be the impact on performance.

If that data and ids are the same, React shouldn't re-render anything. And hooks are stable enough also not to re-fetch data unnecessarily.

Cool, looks like we're aligned!

Reply:
cool, we're using this pattern! (https://github.com/filecoin-project/filecoin-station/pull/123/files#diff-227719cb0a3e80022b8ead60744f738ee9689c17ce6edcc625d427e577b8460dR70)
and we can evolve it to a hook with react context  👍